src/auslib/AUS.py

```diff
@@ -131,16 +133,62 @@ def evaluateRules(self, updateQuery, transaction=None):
         # 3) Incoming release is older than the one in the mapping, defined as one of:
         #    * version decreases
         #    * version is the same and buildID doesn't increase
-        release = releases.get_release(rule["mapping"], transaction, include_sc=False)
-        blob = None
-        if release:
-            blob = createBlob(release["blob"])
-        # TODO: remove me when old releases table dies
-        else:
-            release = dbo.releases.getReleases(name=rule["mapping"], limit=1, transaction=transaction)[0]
-            blob = release["data"]
+        def get_blob(mapping):
+            release = releases.get_release(mapping, transaction, include_sc=False)
+            blob = None
+            if release:
+                blob = createBlob(release["blob"])
+            # TODO: remove me when old releases table dies
+            else:
+                release = dbo.releases.getReleases(name=mapping, limit=1, transaction=transaction)[0]
+                blob = release["data"]
+            return blob
+
+        mapping = rule["mapping"]
+        blob = get_blob(mapping)
         if not blob or not blob.shouldServeUpdate(updateQuery):
             return None, None, eval_metadata
 
-        self.log.debug("Returning release %s", rule["mapping"])
+        version_pin = updateQuery.get("pin")
+        if version_pin is not None:
+            try:
+                version_pin = PinVersion(version_pin)
+            except ValueError:
+                raise BadDataError(f"Version Pin String '{version_pin}' is invalid.")
+            blob_version = blob.getApplicationVersion(updateQuery["buildTarget"], updateQuery["locale"])
+            if blob_version:
+                blob_version = MozillaVersion(blob_version)
+                # We should only pay attention to the pin version if we would otherwise return
+                # something newer. If we unconditionally returned the pinned version, we would do
+                # incorrect things like skipping watersheds.
+                if blob_version > version_pin:
+                    pin_mapping = dbo.pinnable_releases.getPinMapping(updateQuery["product"], getFallbackChannel(updateQuery["channel"]), str(version_pin))
+                    # Note that we fall back to serving the original update if the pin is not found
+                    # in the pin table, even if the version that we will serve is past the pinned
+                    # version. This is because, if there is something wrong with the update pin,
+                    # we would rather default to keeping the user up-to-date.
```

Comment:
Is that an explicit ask from product?  I'd have thought we'd never want to serve a newer version than what was pinned, so this behavior is surprising to me.

Do we know what chrome does there?

I'm a bit confused about the example, too.  How do we (balrog) know 120.15 will never be released, and therefore we point to 134, vs 120.15 is not released yet, therefore we stay on 120.14?

Reply:
> Is that an explicit ask from product?

No. I just think this is definitely the right thing to do here, even if it is a bit counterintuitive.

> How do we (balrog) know 120.15 will never be released

It's not really an issue of knowing whether `120.15` will ever be released. It's an issue of doing the best thing in an ambiguous situation.

This situation is technically unsupported, since part of the requirements for enabling this feature will be to set a pin to an existing version, not a version that they think might exist one day. If they want to pin to the latest version of `120`, that is what a major pin (`120.`) is for. It is not appropriate to guess at what minor version might be desirable before it has even been released.

Despite that, we still have to decide what to do in this unsupported situation. These seem to be the two options:

 1. Treat the invalid pin as invalid and ignore it. If it later becomes valid, anyone that updated in the window when it was invalid will have updated beyond it.
 2. Treat the invalid pin as potentially being valid one day. If it never becomes valid, anyone that didn't update right away will forever be trapped on their current version.

It is my opinion that the first option is both safer (in terms of providing security fixes) and more correct and is therefore the much more preferable option.